imagecraft/pack/chroot.py

```diff
@@ -0,0 +1,189 @@
+from pathlib import Path
+from typing import Any
+
+from craft_cli import emit
+from craft_parts.utils import os_utils
+
+from imagecraft import errors
+
+logger = logging.getLogger(__name__)
+
+
+class Mount:
+    """Mount entry for chroot setup."""
+
+    _fstype: str | None
+    _src: str
+    _relative_mountpoint: str
+    _options: list[str] | None = None
+    _mountpoint: Path | None = None
+
+    def __init__(
+        self,
+        fstype: str | None,
+        src: str,
+        relative_mountpoint: str,
+        *,
+        options: list[str] | None = None,
+        mountpoint: Path | None = None,
+    ) -> None:
+        self._fstype = fstype
+        self._src = src
+        self._relative_mountpoint = relative_mountpoint
+        if options:
+            self._options = options
+        if mountpoint:
+            self._mountpoint = mountpoint
+
+    def mount(self, base_path: Path) -> None:
+        """Mount the mountpoint.
+
+        :param base_path: path to mount the mountpoint under.
+        """
+        args: list[str] = []
+        if self._options:
+            args.extend(self._options)
+        if self._fstype:
+            args.append(f"-t{self._fstype}")
+
+        self._mountpoint = base_path / self._relative_mountpoint.lstrip("/")
+        pid = os.getpid()
+        # Only mount if mountpoint exists.
+        emit.debug(f"Mount {self._src} on {str(self._mountpoint)}")
+        logger.debug("[pid=%d] mount %r on chroot", pid, str(self._mountpoint))
+        if self._mountpoint.exists():
+            logger.debug("[pid=%d] mount %r on chroot", pid, str(self._mountpoint))
+            os_utils.mount(self._src, str(self._mountpoint), *args)
+        else:
+            logger.debug(
+                "[pid=%d] mountpoint %r does not exist", pid, str(self._mountpoint)
+            )
+
+    def umount(self, *, lazy: bool = False) -> None:
+        """Umount the mountpoint."""
+        pid = os.getpid()
+
+        if self._mountpoint and self._mountpoint.exists():
+            logger.debug("[pid=%d] umount: %r", pid, str(self._mountpoint))
+
+            os_utils.mount(str(self._mountpoint), "--make-rprivate")
+            args: list[str] = ["--recursive"]
+
+            # Mount points under /dev may be in use and make the bind mount
+            # unmountable. This may happen in destructive mode depending on
+            # the host environment, so use MNT_DETACH to defer unmounting.
+            if lazy:
+                args.append("--lazy")
+            os_utils.umount(str(self._mountpoint), *args)
+
+
+def _runner(
+    path: Path,
+    conn: Connection,
+    target: Callable[..., str | None],
+    args: tuple[str],
+    kwargs: dict[str, Any],
+) -> None:
+    """Chroot to the execution directory and call the target function."""
+    pid = os.getpid()
+    logger.debug("[pid=%d] child process: target=%r", pid, target)
+    try:
+        logger.debug("[pid=%d] chroot to %r", pid, path)
+        os.chdir(path)
```

Comment:
I am not sure what the significance of changing the directory first is ?

Reply:
It was copied from the original implementation but after testing I cannot determine why that is necessary so I removed it.